File: src/AI/Rystem.PlayFramework/Domain/Models/ClientInteractionResult.cs
using Microsoft.Extensions.AI;

namespace Rystem.PlayFramework;

/// <summary>
/// Result of a client-side tool execution.
/// Client sends this back to server to resume execution.
/// </summary>
public sealed class ClientInteractionResult
{
    /// <summary>
    /// ID of the interaction request this result corresponds to.
    /// Must match the InteractionId from ClientInteractionRequest.
    /// </summary>
    public required string InteractionId { get; init; }

    /// <summary>
    /// Multi-modal contents returned by client tool.
    /// Uses native Microsoft.Extensions.AI types (DataContent, TextContent).
    /// These are directly added to chat history for LLM processing.
    /// </summary>
    public IList<AIContent>? Contents { get; init; }

    /// <summary>
    /// Error message if client tool execution failed.
    /// If set, Contents should be null.
    /// </summary>
    public string? Error { get; init; }

    /// <summary>
    /// Timestamp when client executed the tool.
    /// </summary>
    public DateTimeOffset ExecutedAt { get; init; } = DateTimeOffset.UtcNow;
}

File: src/AI/Rystem.PlayFramework.Client/src/rystem/src/models/ClientInteractionResult.ts
/**
 * AIContent types that can be sent to server.
 * Simplified to 2 types: DataContent (Base64) and TextContent.
 */
export interface AIContent {
    type: "text" | "data";
    text?: string;
    data?: string; // Base64 encoded
    mediaType?: string; // e.g., "image/jpeg", "audio/webm", "application/pdf"
}

/**
 * Result from client-side tool execution.
 * Client sends this back to server to resume execution.
 */
export interface ClientInteractionResult {
    /**
     * Unique interaction ID matching the request.
     */
    interactionId: string;

    /**
     * Multi-modal contents returned by tool.
     * Can include images (Base64), audio, video, files, text.
     */
    contents: AIContent[];

    /**
     * Error message if execution failed.
     */
    error?: string;

    /**
     * Timestamp when tool was executed (ISO 8601).
     */
    executedAt: string;
}
